### PR TITLE
Implement GUI control for placing paralell trajectories with the Leksell frame.

### PR DESCRIPTION
**Pull request by d-vogel**
One of our user-case with the Stereotactic Points module (`stereo_points.py`) is the placement of trajectories when the Ben Gun was used (this is for intra-operative stimulation tests). 

This pull request adds GUI (a comboBox to select one of the 5 possibilities) and an extra transform to shift the trajectory 2mm parallel. For now only the "+ " rotation is supported. Adding the "x" will make the interface heavy, at this point we should probably rethink the GUI deeper.

### CODE REVIEW

**Comment by d-vogel** [comment]
@pieper Up on that ?
any chance I could become a maintainer ?

**Comment by pieper** [comment]
@d-vogel thanks for the PR and the ping.  Your code looks good generally but since I haven't used this module myself I want to be extra careful with merges.  Could you provide some more text (description here and/or comments in the code) about what your change implements and why it is compatible with the previous code? That is, just to clarify that anyone using the previous version won't get unexpected results due to these changes.

File: stereo_points/stereo_points.py
Comment (pieper) [inline stereo_points/stereo_points.py:260]: Stylewise we don't like commented blocks of code like this ("dead code") unless it's there's a clear comment about why keeping it around might be valuable to developers in the future.  Better to get rid of this code if it's really not needed anymore.

**Review by pieper — CHANGES_REQUESTED** [review]
Generally looks good.  Just needs a clearer description of what this change actually implements and what if any other behavior of the code has changed.

**Comment by d-vogel** [comment]
Hey @pieper, I tried to address your comments, thanks for the tips.

**Review by pieper — APPROVED** [review]
I don't see any issues with this and I would be fine with merging it.

But since I don't use this code myself, it would be good if a second set of eyes could review it before we merge.

@lassoan or @JBeninca could you have a look?

File: find_zFrame/find_zFrame.py
Comment (lassoan) [inline find_zFrame/find_zFrame.py:283]: To make the code readable, use a constant variable (`newState == qt.Qt.Checked`) instead of magic number (`newState == 2`).

**Review by lassoan — APPROVED** [review]
The changes look good to me. I've just added a few comments to further improve readability.

File: stereo_points/stereo_points.py
Comment (lassoan) [inline stereo_points/stereo_points.py:216]: Need to store the observer tag that this method returns and when switching to a different `self.coordTableNode` then `RemoveObserver` must be called for the previously observed table.

File: stereo_points/stereo_points.py
Comment (lassoan) [inline stereo_points/stereo_points.py:220]: It is not nice that we take over whatever is the first table in the scene. If we use node references then that should be usable to tell if we have created a table node for this case already. If for some reason we cannot use a reference then at least we should add a node attribute to the tables that this module creates (e.g., `self.coordTableNode.SetAttribute("Stereotaxia", "1")` and we can take the first table node that has that attribute set, not the first node in the scene (that might have been created a different module and may be completely unrelated).

**Review by lassoan — APPROVED** [review]
Thank you, it looks OK, just added two small comments to consider.

**Comment by lassoan** [comment]
Once the added observers are properly removed, this should be ready to merge.

File: stereo_points/stereo_points.py
Comment (d-vogel) [reply stereo_points/stereo_points.py]: Thanks, I've implemented it in [2609bc0](https://github.com/JBeninca/SlicerStereotaxia/pull/9/commits/2609bc03eeb6b2b3209266d4fb710e6355d968e3). I'm not sure if the implementation is correct as I don't really understand why this is needed.

**Review by lassoan — COMMENTED** [review]
It is getting there, but you don't remove all observers that you add and not remove them at the optimal place.

The goal is to not observe some old nodes that you used previously. You only want to get ...Modified methods when the current nodes are modified.

**Comment by d-vogel** [comment]
Thanks for the suggestions.
The thing is, the goal of these observers was actually that they would be active forever after the markupLineNode (here newnode) and TableNode (here coordtable) were created

This is:
1. to keep the names in sync in case the user renames any of the two (since we were using the names to match when, admittedly this is not needed anymore)
2. update the coordTable when the markupLine is moved (by dragging a point in the slice views). This is actually an important user case, when we want to create a trajectory and get the Leksell Arc settings.

So I'm unsure whether we want to ever remove those observers.

**Comment by lassoan** [comment]
When you close the scene and all nodes are deleted then you need to remove the observer. You also need to make sure that everything works well when you load a scene (and the scene anyway contains the line and table that the module created). It is nice if the module behave well when the user deletes one of these nodes.

Making these nodes singletons prevents deletion when the scene is closed and merges singleton nodes when a scene is loaded. An alternative is to use a "parameter" node that stores all numeric parameters and references to the line and table nodes - in this case you can have multiple plans in the scene and choose between them with a node selector.

File: stereo_points/stereo_points.py
Comment (lassoan) [inline stereo_points/stereo_points.py:195]: `type(slicer.vtkMRMLMarkupsLineNode())` is unnecessary and it is an expensive call, as it instantiates an object, just to get its type. You can simply use the type instead:

```suggestion
        if type(newNode) == slicer.vtkMRMLMarkupsLineNode:
```

File: stereo_points/stereo_points.py
Comment (lassoan) [inline stereo_points/stereo_points.py:222]: For each `AddObserver` there must be a matching `RemoveObserver`.
It is great that you save `self.lObs_tableNode`, now you just need to make sure to remove this observer:
- before adding a new observer to another table node (e.g., when you select a different fiducial group)
- in the `cleanup` method (to avoid memory leaks when you exit the application)

To make it easier to remove an observer, I would recommend to save the observed node name as well. For example:

```python
self.tableNodeObservation = [newCoordTable, newCoordTable.AddObserver(vtk.vtkCommand.ModifiedEvent, self.onCoordTableModified)]
```

How to remove the observer:

```python
if self.tableNodeObservation:
    self.tableNodeObservation[0].RemoveObserver(self.tableNodeObservation[1])
    self.tableNodeObservation = None
```

File: stereo_points/stereo_points.py
Comment (lassoan) [inline stereo_points/stereo_points.py:223]: Same applies here as to the table node observation. You have to remove the old observer when the user selects a different fiducial node (and at module cleanup).

File: stereo_points/stereo_points.py
Comment (lassoan) [inline stereo_points/stereo_points.py:599]: Involving SimpleITK to get volume node geometry (origin, spacing, axis directions, and extent) is both unnecessarily complicated (no need to get voxels) and insufficient (because parent transform of the volume is ignored).

You can find code snippet in the script repository that can properly transform between IJK and RAS coordinates (LPS is just inverting the sign of the first two coordinates):

https://slicer.readthedocs.io/en/latest/developer_guide/script_repository.html#get-volume-voxel-coordinates-from-markup-control-point-ras-coordinates

If you want to keep things simpler and do not want to allow positioning the volume using a transform (or you only allow transforming using a linear transform but not a warping transform) then you can check the parent transform and raise an exception.

**Review by lassoan — APPROVED** [review]
I added a few comments about some unnecessarily complex code, missing removal of observations; and the code should be reorganized to separate the widget (GUI manipulations) from logic (computations that you may want to use without a GUI).

However, the proposed changes contain many improvements already, so it does not make sense to delay the merge any further. So, I will merge this pull request as is, and please try to spend some time later to make the suggested improvenents in follow-up pull requests.

File: stereo_points/stereo_points.py
Comment (d-vogel) [reply stereo_points/stereo_points.py:222]: Thanks, this is helpful. We don't want to remove the observer when the user selects another markupline in the combobox. But we still need to save all the observers and cleanup. So I quess tableNodeObservation will be a list to store all observers for cleanup.

**Pull request merged**
